From 32744de9b7c5affbfe115b4c8a724d6efd04caae Mon Sep 17 00:00:00 2001
From: "hertz.wang" <wangh@rock-chips.com>
Date: Wed, 12 Sep 2018 19:00:21 +0800
Subject: [PATCH 1/1] support rk reboot loader

---
 init/halt.c | 8 +++++++-
 init/init.c | 4 ++++
 2 files changed, 11 insertions(+), 1 deletion(-)

diff --git a/init/halt.c b/init/halt.c
index e6dcb1c..4bec394 100644
--- a/init/halt.c
+++ b/init/halt.c
@@ -169,7 +169,7 @@ int halt_main(int argc UNUSED_PARAM, char **argv)
 		RB_POWER_OFF,
 		RB_AUTOBOOT
 	};
-	static const smallint signals[] = { SIGUSR1, SIGUSR2, SIGTERM };
+	static const smallint signals[] = { SIGUSR1, SIGUSR2, SIGTERM, SIGPWR };
 
 	int delay = 0;
 	int which, flags, rc;
@@ -213,6 +213,12 @@ int halt_main(int argc UNUSED_PARAM, char **argv)
 			/* talk to linuxrc */
 			/* bbox init/linuxrc assumed */
 			pid_t *pidlist = find_pid_by_name("linuxrc");
+			//set reboot parameter
+			if (signals[which] == SIGTERM) {
+				if (argc > 1 && strstr(argv[1], "loader") != NULL) {
+					which++;
+				}
+			}
 			if (pidlist[0] > 0)
 				rc = kill(pidlist[0], signals[which]);
 			if (ENABLE_FEATURE_CLEAN_UP)
diff --git a/init/init.c b/init/init.c
index 6439e2b..96f70f1 100644
--- a/init/init.c
+++ b/init/init.c
@@ -813,6 +813,7 @@ static void run_shutdown_and_kill_processes(void)
  * and only one will be remembered and acted upon.
  */
 
+#define RB_AUTOBOOT_CMD 0xa1b2c3d4
 /* The SIGPWR/SIGUSR[12]/SIGTERM handler */
 static void halt_reboot_pwoff(int sig) NORETURN;
 static void halt_reboot_pwoff(int sig)
@@ -838,6 +839,9 @@ static void halt_reboot_pwoff(int sig)
 	} else if (sig == SIGUSR2) {
 		m = "poweroff";
 		rb = RB_POWER_OFF;
+	} else if (sig == SIGPWR) {
+		m = "reboot loader";
+		rb = RB_AUTOBOOT_CMD;
 	}
 	message(L_CONSOLE, "Requesting system %s", m);
 	pause_and_low_level_reboot(rb);
-- 
2.7.4

